### PR TITLE
UTOPIA-211-add http health check endpoint

### PR DESCRIPTION
**Pull request by binggaofw**
# Description

This PR includes the following proposed change(s):

- {List all the changes, if possible add the jira ticket #}

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Development Dependency Working Agreement
- [ ] My code DOES NOT include the importing of new dependencies into the DPIA ecosystem
- [ ] My code DOES include the importing of new dependencies into the DPIA ecosystem
**If new dependencies are being introduced to the DPIA ecosystem:**
- [ ] The functionality of the dependency drastically reduces code complexity and makes my changes more easily maintainable and readible 
- [ ] The dependency being introduced does not contain multiple layers of nested dependencies introducing maintainability complexity to the DPIA ecosystem

## Frontend Development Changes
- [ ] N/A
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to project documentation or diagrams that reflect my changes
- [ ] New and existing unit tests pass locally with my changes
- [ ] My code follows Airbnb React Style Guidelines

## API Development Changes
- [ ] N/A
- [ x] I have performed a self-review of my own code
- [ ] My code follows standards and practices outlined in the BC Government API Development Guidelines
- [ ] New and existing unit tests pass locally with my changes
- [ ] My changes includes Swagger documentation updates that reflect the changes I am introducing

### CODE REVIEW

File: src/backend/package.json [hunk 1]
```diff
@@ -21,9 +21,11 @@
     "test:e2e": "jest --config ./test/jest-e2e.json"
   },
   "dependencies": {
+    "@nestjs/axios": "0.1.0",
```
Comment (antsand) [inline src/backend/package.json:24]: can we stick to pure javascript async and await. Axios is not really adding anything much to async/await.

Also there are multiple nested sub dependencies which will cause trouble during npm updates in the future.

File: src/backend/package.json [hunk 2]
```diff
@@ -21,9 +21,11 @@
...
     "@nestjs/common": "9.0.0",
     "@nestjs/core": "9.0.0",
     "@nestjs/platform-express": "9.0.0",
+    "@nestjs/terminus": "9.1.2",
```
Comment (antsand) [inline src/backend/package.json:28]: Is it necessary to have terminus? Can we keep the default nestjs package.json without adding more to this ginormous list?

File: src/backend/src/health/health.controller.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  constructor(
+    private health: HealthCheckService,
+    private http: HttpHealthIndicator,
+  ) {}
+
+  @Get()
+  @HealthCheck()
+  check() {
+    return this.health.check([
+      () => this.http.pingCheck('dpia-web', 'http://localhost:3000/api'),
```
Comment (kushal-arora-fw) [inline src/backend/src/health/health.controller.ts:20]: Sorry, I might be new on this `Terminus` usage. But looks like from the documentation from [here](https://docs.nestjs.com/recipes/terminus) - the purpose of Terminus is to do health/ping checks on an external site. Aren't we here pinging our own servers continuously consuming resources and memory?  I think this is something openshift should take care and not every deployed node instance.

Please let me know if my understanding is different. Happy to connect over call.

isn't the purpose of pingCheck /

File: src/backend/package.json (hunk 2 above)
Comment (binggaofw) [reply src/backend/package.json:28]: it depends on what health check we want to do in the future, NestJS provide many different indicators to monitor our system, like database, CPU, mem, etc even we can define our custom health check indicator. so yes, if only a simple http health check, we do not need terminus, but if we do want to implement a sophisticated HC, I think we need this tool.

Comment (kushal-arora-fw) [reply src/backend/package.json:28]: I think we should delegate all the health monitoring related heavy lifting to Openshift.

There can be multiple node instances in future - we do not want all the nodes to take care of these types of monitoring using resources. What do you think?

File: src/backend/package.json (hunk 1 above)
Comment (binggaofw) [reply src/backend/package.json:24]: i will check this

File: src/backend/src/health/health.controller.ts (hunk 3 above)
Comment (binggaofw) [reply src/backend/src/health/health.controller.ts:20]: Then I think we can delete this package, just do a simple http check for hc endpoint

Comment (akroon3r) [reply src/backend/src/health/health.controller.ts:20]: yea all we need to do is expose a healthcheck endpoint to serve back a 200 status. Openshift deployment configs allow for the addition of readiness and liveness probes that will target this endpoint to determine whether or not the container is healthy.

File: src/backend/src/health/health.controller.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import { ApiExcludeEndpoint } from '@nestjs/swagger';
+@Controller('health')
+export class HealthController {
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  constructor() {}
+
+  @Get()
+  @HealthCheck()
+  @ApiExcludeEndpoint()
+  async check() {
```
Comment (kushal-arora-fw) [inline src/backend/src/health/health.controller.ts:16]: Can you please add Swagger Annotations and return type here?

File: src/backend/src/main.ts [hunk 5]
```diff
@@ -3,11 +3,12 @@ import 'dotenv/config';
...
 
 async function bootstrap() {
+
   // Initialize the Express based Nest application
   const { app } = await createNestApp();
   await app.init();
 
   // App listen to a port
-  await app.listen(process.env.PORT);
+  await app.listen(process.env.PORT || 3000);
```
Comment (kushal-arora-fw) [inline src/backend/src/main.ts:12]: I would request you to add `.env` to your `backend` and get rid of `OR` statement here

```
ENV = development
PORT = 3000

API_NAME = Digital Privacy Impact Assessment API
API_VERSION = 0.0.1
```

File: src/backend/src/main.ts [hunk 6]
```diff
@@ -3,11 +3,12 @@ import 'dotenv/config';
 import { createNestApp } from './app.config';
 
 async function bootstrap() {
+
```
Comment (kushal-arora-fw) [inline src/backend/src/main.ts:6]: unnecessary change

File: src/backend/src/health/health.controller.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import {
+  HealthCheckService,
+  HttpHealthIndicator,
+  HealthCheck,
+} from '@nestjs/terminus';
+import { ApiExcludeEndpoint } from '@nestjs/swagger';
+@Controller('health')
+export class HealthController {
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  constructor() {}
```
Comment (kushal-arora-fw) [inline src/backend/src/health/health.controller.ts:11]: I strongly urge not to disable eslint unless unavoidable If it is an empty constructor, I guess you can get rid of it.

File: src/backend/src/health/health.controller.spec.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    const module: TestingModule = await Test.createTestingModule({
+      controllers: [HealthController],
+    }).compile();
+
+    controller = module.get<HealthController>(HealthController);
+  });
+
+  it('should be defined', () => {
+    expect(controller).toBeDefined();
+  });
```
Comment (kushal-arora-fw) [inline src/backend/src/health/health.controller.spec.ts:17]: You can also check if your function returns `HttpStatus.OK` if that's what you're returning??

**Review by kushal-arora-fw — CHANGES_REQUESTED** [review]

File: src/backend/src/health/health.controller.ts (hunk 7 above)
Comment (binggaofw) [reply src/backend/src/health/health.controller.ts:11]: but this is an empty constructor, that's why I add it

Comment (kushal-arora-fw) [reply src/backend/src/health/health.controller.ts:11]: I mean - you can get rid of the constructor itself.

File: src/backend/src/main.ts (hunk 6 above)
Comment (binggaofw) [reply src/backend/src/main.ts:6]: good catch, remove it

File: src/backend/src/main.ts (hunk 5 above)
Comment (binggaofw) [reply src/backend/src/main.ts:12]: done

File: src/backend/src/health/health.controller.ts (hunk 4 above)
Comment (binggaofw) [reply src/backend/src/health/health.controller.ts:16]: I thought may be this health check do not need to expose, but sure, I can remove  ApiExcludeEndpoint() and add it to swagger

File: src/backend/src/health/health.controller.spec.ts (hunk 8 above)
Comment (binggaofw) [reply src/backend/src/health/health.controller.spec.ts:17]: yes. i do not change the default test file yet. but i will change it

**Comment by kushal-arora-fw** [comment]
Also, while you're at it @binggaofw - can you also remove `package-lock.js` file as well from the change as well (and ideally from your system as well). It will have the unnecessary changes like Terminus and Axios and their dependencies.

**Review by kushal-arora-fw — APPROVED** [review]
Great work! Looks good 👌

**Review by antsand — APPROVED** [review]
Looks good!
